prototype_source/maskedtensor_adagrad.py

```diff
@@ -0,0 +1,208 @@
+# -*- coding: utf-8 -*-
+
+"""
+(Prototype) Efficiently writing "sparse" semantics for Adagrad with MaskedTensor
+================================================================================
+"""
+
+######################################################################
+# Before working through this tutorial, please review the MaskedTensor
+# `Overview <https://pytorch.org/tutorials/prototype/maskedtensor_overview.html>`__ and
+# `Sparsity <https://pytorch.org/tutorials/prototype/maskedtensor_sparsity.html>`__ tutorials.
+#
+# Introduction and Motivation
+# ---------------------------
+# `Issue 1369 <https://github.com/pytorch/pytorch/issues/1369>`__ discussed the additional lines of code
+# that were introduced while writing "sparse" semantics for Adagrad.
+# But really, the code uses sparsity as a proxy for masked semantics rather than the intended use case of sparsity:
+# a compression and optimization technique,
+# Previously, we worked around the lack of formal masked semantics by introducing one-off semantics and operators
+# while forcing users to be aware of storage details such as indices and values.
+#
+# In particular, we'll point out when sparsity is used as a semantic extension, i.e. unspecified values are not zero
+# and when it is just used to compress zeros.
```

Comment:
Is the diff between sparse vs masked that sparse is usually used to compress 0s and masked is just a way for us to "ignore" certain values? From this sentence, it seems to say that sparse semantics = there are nonzero unspecified values, but masked = the only unspecified values are 0s...which seems opposite of what I originally thought.

Reply:
I see, it's meant to say that sparsity is used to convey masked semantics (i.e. semantic extension).. and technically sparsity can be used to compress other values as well, so I'll change this